Entry Controller

senvarin opened the pull request.

senvarin committed:
Entry Controller

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("entries-controller")
```
Обычно не добавляют суффикс `-controller` в api. Лучше оставить только название сущности.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @PostMapping("creat")
+    public void creat(@RequestBody EntryDTO entryDTO) {
+        service.creat(mapper.toModel(entryDTO));
```
То, что было создано, стоит вернуть в ответ.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+    private final EntryService service;
+    private final EntryMapper mapper;
+
+    @PostMapping("creat")
```
Статус успеха должен быть 201 (CREATED)

snusmum replied:
Буква потерялась. И так на всём дереве вызовов.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+        return mapper.toDTO(service.getExisting(id));
+    }
+
+    @GetMapping("search-name/{name}")
```
Должен быть один хвост `/all`, принимающий опциональный параметр (не часть пути) `name` для поиска.

senvarin replied:
Я правильно понял, что метод `searchByName` нужно убрать и сделать чтобы у метода `getAll` был необязательный параметр **name**?

snusmum replied:
Да.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+    @PostMapping("{id}/update")
+    public void updateEntry(@PathVariable Long id,
+                             @RequestBody EntryDTO entry) {
+        service.update(id, mapper.toModel(entry));
```
Тоже можно вернуть.

snusmum commented on pom.xml:
```diff
@@ -14,14 +14,88 @@
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <source>${java.version}</source>
+          <target>${java.version}</target>
```
С какой версией Java в итоге собирается приложение?

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/dto/EntryDTO.java:
```diff
@@ -0,0 +1,15 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class EntryDTO {
```
В зависимостях сваггер подключен, а в дто документации нет.

snusmum replied:
Поля тоже принято описывать. Как минимум те, что будут отдаваться на фронт.
В них отмечаются те, что обязательно вернутся | обязательны для заполнения.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+    private final EntryMapper mapper;
+
+    @PostMapping("creat")
+    public void creat(@RequestBody EntryDTO entryDTO) {
```
Для параметров каждого метода (создание, обновление) нужно использовать отдельный класс, даже если есть совпадение по полям.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @GetMapping("all")
+    public List<EntryDTO> getAll() {
```
В таком варианте на фронте не узнают идентификатор, по которому удалять или обновлять запись.
Он должен быть в DTO.

snusmum commented on src/main/java/org/kaoden/ws/homework/repository/EntryRepositoryImp.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public void creat(Entry entry) {
+        Long id = getFreeId();
```
После получения идентификатора в сервисе он будет заново вычислен ещё раз?

snusmum commented on src/main/java/org/kaoden/ws/homework/repository/EntryRepositoryImp.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public Long getFreeId() {
```
Запускаемый сервер умеет обрабатывать несколько потоков одновременно. Если не синхронизироваться, то можно получить интересные эффекты. Но это пока просто для информации.

snusmum commented on src/test/java/org/kaoden/ws/homework/AppTest.java:
А зачем удалять? Обычно его оставляют как индикатор того, что ничего не сломано и контекст поднимается.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntryControllerIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+    MockMvc mockMvc;
+
+    @MockBean
+    EntryRepository repository;
```
В интеграционном тесте не надо мокать репозиторий.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntryControllerIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+                        .contentType(APPLICATION_JSON_UTF8)
+                        .content(OW.writeValueAsString(getEntryDto("Test"))))
+                .andDo(print())
+                .andExpect(status().isOk());
```
Если репозиторий замокан, то нужно проверять, что его соответствующий метод вызван с правильными аргументами.
Когда мок будет убран, то нужно будет проверять, что в репозитории появилась ожидаемая запись.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntryControllerIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+    static final String URL = "/entries-controller/";
+    static final MediaType APPLICATION_JSON_UTF8 = new MediaType(MediaType.APPLICATION_JSON.getType(), MediaType.APPLICATION_JSON.getSubtype(), StandardCharsets.UTF_8);
+    static final ObjectWriter OW = new ObjectMapper().writer();
```
Не стоит экономить буквы на переменные.
А для маппинга и самого ObjectMapper с методом writeValueAsString.

snusmum commented on src/test/java/org/kaoden/ws/homework/repository/EntryRepositoryImpTest.java:
```diff
@@ -0,0 +1,69 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
```
Эта аннотация поднимает контекст. В текущей ситуации это излишне.

snusmum commented on src/test/java/org/kaoden/ws/homework/repository/EntryRepositoryImpTest.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Test
+    void creatNull() {
+        assertThrows(NullPointerException.class, () ->  repository.creat(null));
```
Странные ожидания. Стоило добавить @NonNull и проверять, что сообщение об NPE будет осмысленным.

senvarin replied:
Пока не стал исправлять это. Сделаю уже в следующей домашек

snusmum commented on src/test/java/org/kaoden/ws/homework/repository/EntryRepositoryImpTest.java:
```diff
@@ -0,0 +1,69 @@
+class EntryRepositoryImpTest {
+
+    @Autowired
+    EntryRepository repository;
```
Заполнения нет, в итоге в тестах проверяется только поведение, когда данные не найдены. А зачем?

snusmum commented on src/test/java/org/kaoden/ws/homework/repository/EntryRepositoryImpTest.java:
Где-то потерялись тесты сервиса.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntryControllerIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@FieldDefaults(level = AccessLevel.PRIVATE)
+class EntryControllerIntegrationTest {
```
Названия классов с интеграционными обычно заканчиваются на IT, чтобы можно было отделить от юнит-тестов при помощи maven-плагина и запускать по отдельности.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntryControllerIntegrationTest.java:
```diff
@@ -0,0 +1,143 @@
+
+    @Test
+    void createEntryInRep() throws Exception {
+        mockMvc.perform(post(URL + "creat")
```
nit: Если не хочется каждый раз писать url, то базовую часть лучше указывать через параметр, а не собирать через конкатенацию.
`post("/{url}/create", URL)`

snusmum commented on src/main/java/org/kaoden/ws/homework/model/Entry.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+@Builder
+@FieldDefaults(level = PRIVATE)
+public class Entry {
```
nit: Элемент в классе Map тоже Entry. Возможно стоит выбрать какое-то более "бизнесовое" название.

snusmum reviewed (commented):
Название приложения rest-controller как-то не отражает его назначения. Зачастую они называются по основной сущности, ради работы с которой созданы (complaint-service), либо по выполняемому функционалу (notification-service)

senvarin committed:
Refactoring

URL of api was changed to /entries/
Search by name method was deleted and "getAll" method mow contain optional
parametr "name" for searching by name
CreatEntryDto and UpdateEntryDto were added as creating and updating
Dto
Also were fixed all test and added for Service and App1

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,59 @@
+        if (name == null)
+            return mapper.toDTOList(service.getAll());
+        else
+            return mapper.toDTOList(service.search(name));
```
Обычно такую логику на уровне сервиса реализуют.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/EntryController.java:
```diff
@@ -0,0 +1,59 @@
+            return mapper.toDTOList(service.search(name));
+    }
+
+    @GetMapping("search-id/{id}")
```
Просто `{id}` достаточно для хвоста получения по идентификатору

snusmum commented on src/main/java/org/kaoden/ws/homework/repository/EntryRepositoryImp.java:
```diff
@@ -0,0 +1,72 @@
+        if (entryExist(id))
+            return entriesStorage.get(id);
+        else
+            throw new NotFoundException("There is no entry with this ID: " + id);
```
Эти проверки разумнее переместить в сервис. Можно в следующем ПР.

snusmum replied:
Если if-else с переносами для каждого кейса, то принято ставить скобки.

snusmum commented on src/test/java/org/kaoden/ws/homework/AppTest.java:
```diff
@@ -1,38 +1,16 @@
-        assertTrue( true );
+    @Test
+    void contextLoads(ApplicationContext context) {
+        assertThat(context).isNotNull();
```
Имелось ввиду, что пустой тест самим фактом успешного запуска проверит, что нет ошибок в конфигурировании бинов или ещё чего-то.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,167 @@
+                    .id(0L)
+                    .name("test")
+                    .build();
+    }
```
Это обычно оформляют в виде переменных. Методы используют, когда удобно несколькими параметрами создать данные для очередного теста, а всё остальное - общая логика, которую нет смысла копипастить.

snusmum replied:
В тестах нужно передавать|принимать максимум возможных данных. В данном случае не хватает description, link.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,167 @@
+    void getAllEntries() throws Exception {
+        mockMvc.perform(get("/{url}/all", URL))
+               .andDo(print())
+               .andExpect(status().isOk());
```
А в чём состоит тест? Нужно заполнить репозиторий несколькими записями (как минимум две должны подходить под критерии поиска, как минимум одна  - нет). Дальше проверить, что всё вернётся.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,167 @@
+
+        // Assert
+        assertThat(result.getResponse().getContentAsString())
+                .isEqualTo(expectedEntry);
```
Нерабочий тест, т.к. заполнение дто обновления совпадает с тем, что в лежащей в репозитории карточке.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+class EntryServiceImpTest {
+
+    @Mock
+    static EntryRepository repository = new EntryRepositoryImp();
```
Не нужно создавать. Mock сделает это.

snusmum replied:
не нужно делать поле статическим

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Mock
+    static EntryRepository repository = new EntryRepositoryImp();
+    static EntryService service;
```
`@InjectMock` над сервисом и не нужно будет руками создавать сервис в методе ниже.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+        return EntryArgument.builder()
+                            .name("Test")
+                            .build();
+    }
```
Все поля должны заполняться.

snusmum replied:
Если это нужно ровно для одного теста, то в том тесте лучше и оставить. Станет актуально, когда аргументы создания и обновления разделятся.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+    void createEntry() {
+        // Arrange
+        Entry expectedEntry = getEntry();
+        when(repository.create(expectedEntry)).thenReturn(expectedEntry);
```
Возвращать лучше другую сущность. Для неё можно мок создать.
Иначе придётся ещё проверять, что метод репозитория был вызван.

snusmum commented on src/main/java/org/kaoden/ws/homework/service/argument/EntryArgument.java:
```diff
@@ -0,0 +1,15 @@
+    String name;
+    String description;
+    String link;
+}
```
Та же история, что с  дто. Каждому методу свой аргумент.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+    void deleteEntry() {
+        // Arrange
+        Long id = 0L;
+        doNothing().when(repository).delete(id);
```
И без этого должно работать.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,127 @@
+        service.delete(id);
+
+        // Assert
+        verify(repository, times(1)).delete(id);
```
По умолчанию проверяется, что вызов ровно 1. Тоже лучше не писать лишнее.

senvarin committed:
Minor code refactoring

senvarin committed:
Fix test

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/dto/EntryDTO.java:
```diff
@@ -0,0 +1,24 @@
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@Schema(description = "DTO to return Entry")
+public class EntryDTO {
+    @Schema(name = "Entry ID in storage")
```
Идентификатора не может не быть.

snusmum commented on src/main/java/org/kaoden/ws/homework/controller/dto/EntryDTO.java:
```diff
@@ -0,0 +1,24 @@
+    Long id;
+    @Schema(name = "Entry name", requiredMode = REQUIRED)
+    String name;
+    @Schema(name = "Entry description", requiredMode = REQUIRED)
```
А вот описание вряд ли так жизненно необходимо, равно как и ссылка. Например запись о книге может обойтись и без этой информации.

snusmum commented on src/main/java/org/kaoden/ws/homework/service/EntryServiceImp.java:
```diff
@@ -0,0 +1,55 @@
+        if (searchText == null)
+            return repository.getAll();
+        else
+            return repository.findByName(searchText);
```
Здесь тернарка была бы уместна `return <condition> ? <expression when true> : <expression when false>`
Но если без неё, то см. коммент, оставленный выше про скобки.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,189 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@FieldDefaults(level = AccessLevel.PRIVATE)
+class EntriesServiceIT {
```
Это тест контроллера

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,189 @@
+        // Arrange
+        Long id = 0L;
+        String name = "test";
+//        String testEntry = WRITER.writeValueAsString();
```
Лучше сразу удалять ненужное.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,189 @@
+                                                            .name("test-update")
+                                                            .description("test-description")
+                                                            .link("test-link")
+                                                            .build();
```
Эти переменные каждая в своём единственном тесте используются. Лучше там и разместить.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,189 @@
+        String name2 = "test-2";
+        repository.create(getEntry(0L, name1));
+        repository.create(getEntry(1L, name2));
+        String expectedList = WRITER.writeValueAsString(Lists.newArrayList(getEntryDto(0L, name1), getEntryDto(1L, name2)));
```
По паре раз вызваны методы для создания одинаковых сущностей. Можно было переменные завести.

senvarin replied:
В прошлый раз вы говорили, что в тетстах должно использоваться по максимуму данных, а поскольку в этом тесте мне по факту нужны разные `id` и `name`, для наглядности, того что они разные я не стал создавать полноценные `Entry `, как локальные переменные и заполнять их полностью, ведь для этого уже есть метод.

snusmum replied:
Выше в метод create репозитория для заполнения передаётся результат вызова `getEntry(0L, name1)` и тот же метод вызывается при заполнении списка ожидаемых значений.

В комментарии предлагается завести локальную переменную `entry1 = getEntry(0L, name1)` и использовать в обоих случаях. Также и для второго.

Не принципиально. Можно оставить и так.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesServiceIT.java:
```diff
@@ -0,0 +1,189 @@
+        Long id = 0L;
+        String updateEntry = WRITER.writeValueAsString(testUpdateEntryDto);
+        String expectedEntry = WRITER.writeValueAsString(getEntryDto(id, "test-update"));
+        repository.create(getEntry(id, "test"));
```
По - прежнему обновляется только одно поле. Можно пойти и убрать код, копирующий описание и ссылку, а тест продолжит успешно проходить.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,134 @@
+                                                          .description("test-description")
+                                                          .link("test-link")
+                                                          .build();
+        when(repository.create(testEntry)).thenReturn(testEntry);
```
Сейчас можно перестать вызывать метод репозитория и тест продолжит проходить. Нужно либо возвращать другую сущность, либо verify-ить вызов.

senvarin replied:
У меня упал тест после удаление этой строчки
![image](https://github.com/KisiMissi/rest-controller/assets/77132667/58b88251-bb81-4712-97a3-c631731e7630)

snusmum replied:
Криво выразился. Нужно ещё начать возвращать свежесозданную запись. Т.е. можно поменять код так, что данные перестанут сохраняться, но при этом тест продолжит проходить.

snusmum commented on src/test/java/org/kaoden/ws/homework/repository/EntryRepositoryImpTest.java:
```diff
@@ -0,0 +1,114 @@
+        // Assert
+        assertThat(actualEntry).isEqualTo(expectedEntry);
+    }
+
```
Потерялись тесты получения списка и поиска по наименованию.

senvarin committed:
Minor code rafactoring

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesControllerIT.java:
```diff
@@ -0,0 +1,194 @@
+
+
+        // Act
+        MvcResult result = mockMvc.perform(get("/{url}/all?name={name}", URL, name2))
```
В таком виде тест показывает, что запись точно найдётся, когда совпадает наименование. Но он не показывает, что поиск работает по вхождению строки без учёта регистра.

Проще всего этого добиться, добавив третью запись с наименованием, например, "some NaMe21"

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesControllerIT.java:
```diff
@@ -0,0 +1,194 @@
+        String name2 = "test-2";
+        repository.create(getEntry(0L, name1));
+        repository.create(getEntry(1L, name2));
+        String expectedList = WRITER.writeValueAsString(Collections.singletonList(getEntryDto(1L, name2)));
```
Если где-то в ПР оставляются комментарии и в результате выполняются правки, то применять их следует во всех подобных случаях, а не только в том месте, где оставлен коммент.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,140 @@
+                           .name("test")
+                           .description("test-description")
+                           .link("test-link")
+                           .build();
```
Это мог быть мок.

senvarin replied:
Никак не могу понять, что вы имеете в виду. Подразумевается объявление `Entry `с аннотацией `@Mock`? И если так, то почему следует сделать так?

snusmum replied:
Или `Entry entry = Mockito.mock(Entry.class)`;
Это позволяет дополнительно проверить, что после возвращения из репозитория не было попыток модифицировать его: `verifyZeroInteractions(entry)`;
Кроме того, в полноценном объекте здесь нет необходимости.

snusmum commented on src/test/java/org/kaoden/ws/homework/service/EntryServiceImpTest.java:
```diff
@@ -0,0 +1,140 @@
+        Entry actualEntry = service.create(argument);
+
+        // Assert
+        assertThat(actualEntry).isEqualTo(testEntry);
```
И сравнивать нужно с тем, что вернул репозиторий, т.к. тест проверяет, что 
- в репозиторий положили ровно то, что надо (проверка обеспечивается отработкой условия when)
- отдали то, что вернул репозиторий

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesControllerIT.java:
```diff
@@ -0,0 +1,194 @@
+        Entry entry2 = getEntry(1L, name2);
+        repository.create(entry1);
+        repository.create(entry2);
+        String expectedList = WRITER.writeValueAsString(Lists.newArrayList(entry1, entry2));
```
`List.of`, чтобы не использовать AssertJ-евские методы там, где есть стандартные аналоги.

snusmum commented on src/test/java/org/kaoden/ws/homework/controller/EntriesControllerIT.java:
```diff
@@ -0,0 +1,194 @@
+@SpringBootTest
+@AutoConfigureMockMvc
+@FieldDefaults(level = AccessLevel.PRIVATE)
+class EntriesControllerIT {
```
Класс должен называться EntrieControllerIT, чтобы Idea его находила по Ctrl+Shift+T.

snusmum reviewed (approved):
В репозитории нет файла .gitignore

senvarin committed:
Fixed tests
